api/v1/broker.go

```diff
@@ -1,35 +1,171 @@
+				ProviderTenantState:     string(sp.TenantSpState),
+				ProviderMetadata:        sp.TenantSpsMeta,
+				ProviderTenantStateInfo: sp.TenantSpStateInfo,
+				AttributeValues:         make(map[string]string),
+			}
+
+			// Construct the attribute values for this SP
+			for _, attribute := range sp.SPAttributes {
+				candidateSP.AttributeValues[strconv.Itoa(int(attribute.AttributeLabelID))] = strconv.Itoa(int(attribute.AttributeValueID))
+			}
+
+			payload.CandidateSPs = append(payload.CandidateSPs, candidateSP)
+		}
+
+		br = append(br, payload)
+	}
+
+	return c.JSON(http.StatusOK, CreateSuccessResponseEnvelope(c, br))
+}
+
+/** example payload
+ * [
+ *   { "spid" : 1 , tenantID: "5", "storage_contract_cid": "bafy1234", "authorization": "abcd" }
+ * ]
+ */
+type BrokerSubscriptionEvents []SubscriptionEventPayload
+
+type SubscriptionEventPayload struct {
+	SPID               db.ID  `json:"sp_id"`
+	TenantID           db.ID  `json:"tenant_id"`
+	StorageContractCID string `json:"storage_contract_cid"`
+	Authorization      string `json:"authorization"`
+}
+
+// /broker/subscription-event [post]
+func (a *apiV1) handlePostBrokerSubscriptionEvent(c echo.Context) error {
+	var subscriptionEvents BrokerSubscriptionEvents
+
+	err := json.NewDecoder(c.Request().Body).Decode(&subscriptionEvents)
+
+	if err != nil {
+		return c.JSON(http.StatusInternalServerError, CreateErrorResponseEnvelope(c, http.StatusInternalServerError, err.Error()))
+	}
+
+	err = a.db.Transaction(func(tx *gorm.DB) error {
+		for _, se := range subscriptionEvents {
+
+			tenant := db.Tenant{}
+			res := tx.Model(&db.Tenant{}).Where("tenant_id = ?", se.TenantID).First(&tenant)
+			if res.Error != nil {
+				// Tenant ID does not exist - this should not happen and represents invalid input
+				return fmt.Errorf("unable to find tenant with id " + strconv.Itoa(int(se.TenantID)))
+			}
+
+			var tenantSettings db.TenantSettings
+			var nextState db.TenantSpState = db.TenantSpStatePending
+
+			err := tenant.TenantSettings.AssignTo(&tenantSettings)
+			if err != nil {
+				// unable to unmarshal tenant settings json - fallback to Pending state
+				log.Error("unable to extract tenant settings from db: " + err.Error())
+			} else {
+				if tenantSettings.SpAutoApprove {
+					nextState = db.TenantSpStateActive
+				} else {
+					nextState = db.TenantSpStatePending
+				}
+			}
+
+			meta := db.TenantsSpsMeta{
+				Signature:         se.Authorization,
+				SignedContractCID: se.StorageContractCID,
+			}
+			var pgMeta pgtype.JSONB
+			pgMeta.Set(meta)
+
+			var tsp db.TenantsSPs
+			res = tx.Model(&db.TenantsSPs{}).Where("tenant_id = ? AND sp_id = ?", se.TenantID, se.SPID).First(&tsp)
+			if res.Error != nil {
+				// Tenant-SP mapping does not exist - represents invalid input
+				return fmt.Errorf("unable to find tenant sp with tenant id " + strconv.Itoa(int(se.TenantID)) + " and sp id " + strconv.Itoa(int(se.SPID)))
+			}
+
+			tsp.TenantSpState = nextState
```

Comment:
Any reason to not create the tsp with these values already set?

Reply:
because we define `tsp` and then give it to the gorm `.First()` function to read the value from the database and marshal it into `tsp` - which could overwrite the values if we initialize with with them already present. This way we can be sure it will not overwrite as we set them after